libraries/AP_InertialSensor/AP_InertialSensor.cpp

```diff
@@ -1933,13 +1933,25 @@ void AP_InertialSensor::update(void)
         // set primary to first healthy accel and gyro
         for (uint8_t i=0; i<INS_MAX_INSTANCES; i++) {
             if (_gyro_healthy[i] && _use(i)) {
-                _primary_gyro = i;
+                _first_healthy_gyro = i;
+#if AP_AHRS_ENABLED
+                // ask AHRS for the true primary, might just be us though
+                _primary_gyro = AP::ahrs().get_primary_gyro_index();
```

Comment:
this could flip-flop between two IMUs, having AP_InertialSensor call into AHRS is dangerous, as AHRS itself depends on AP_InertialSensor, this is a nasty layering violation